src/java.base/share/classes/java/lang/template/TemplateRuntime.java

```diff
@@ -0,0 +1,276 @@
+import java.lang.reflect.Field;
+import java.lang.reflect.Modifier;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+
+import jdk.internal.access.JavaLangAccess;
+import jdk.internal.access.JavaLangInvokeAccess;
+import jdk.internal.access.JavaUtilCollectionAccess;
+import jdk.internal.access.SharedSecrets;
+import jdk.internal.javac.PreviewFeature;
+
+/**
+ * This class provides runtime support for string templates. The methods within
+ * are intended for internal use only.
+ *
+ * @since 20
+ */
+@PreviewFeature(feature=PreviewFeature.Feature.STRING_TEMPLATES)
+final class TemplateRuntime {
+
+    /**
+     * {@link MethodHandle} to {@link TemplateRuntime#getValue(int, StringTemplate)}.
+     */
+    private static final MethodHandle GET_VALUE_MH;
+
+    /**
+     * Initialize {@link MethodHandle MethodHandles}.
+     */
+    static {
+        try {
+            MethodHandles.Lookup lookup = MethodHandles.lookup();
+
+            MethodType mt = MethodType.methodType(Object.class, int.class, StringTemplate.class);
+            GET_VALUE_MH = lookup.findStatic(TemplateRuntime.class, "getValue", mt);
+        } catch (ReflectiveOperationException | SecurityException ex) {
+            throw new AssertionError("template runtime init fail", ex);
+        }
+    }
+
+    /**
+     * Private constructor.
+     */
+    private TemplateRuntime() {
+        throw new AssertionError("private constructor");
+    }
+
+    private static final JavaLangAccess JLA = SharedSecrets.getJavaLangAccess();
+
+    private static final JavaLangInvokeAccess JLIA = SharedSecrets.getJavaLangInvokeAccess();
+
+    private static final JavaUtilCollectionAccess JUCA = SharedSecrets.getJavaUtilCollectionAccess();
+
+    /**
+     * Collect nullable elements from an array into a unmodifiable list.
+     *
+     * @param elements  elements to place in list
+     *
+     * @return unmodifiable list.
+     *
+     * @param <E>  type of elements
+     */
+    @SafeVarargs
+    @SuppressWarnings("unchecked")
+    private static <E> List<E> toList(E... elements) {
+        return JUCA.listFromTrustedArrayNullsAllowed(elements);
+    }
+
+    /**
+     * Return the types of a {@link StringTemplate StringTemplate's} values.
+     *
+     * @param st  StringTemplate to examine
+     *
+     * @return list of value types
+     *
+     * @throws NullPointerException if st is null
+     *
+     * @implNote The default method determines if the {@link StringTemplate}
+     * was synthesized by the compiler, then the types are precisely those of the
+     * embedded expressions, otherwise this method returns the values list types.
+     */
+    static List<Class<?>> valueTypes(StringTemplate st) {
+        Objects.requireNonNull(st, "st must not be null");
+        List<Class<?>> result = new ArrayList<>();
+        Class<?> tsClass = st.getClass();
+        if (tsClass.isSynthetic()) {
+            try {
```

Comment:
I do not know if this code is worth of optimizing but the way to avoid to recompute the List<Class<?>> each time is to use a java.lang.ClassValue and store the classes inside an unmodifiable List. (Field[] -> Class<?>[] -> List<Class<?>>) The last leg can be done just by calling List.of(), there is no need for an ArrayList here

Reply:
Will use List.of. I think use case is raw and caching should be left to the user.

Reply:
i agree

Reply:
I agree. Need a balance of performance and #classes. Just something I'll work on during preview.

Reply:
Changing